zcash_client_sqlite/src/for_async/mod.rs

```diff
@@ -0,0 +1,488 @@
+    ) -> Result<Vec<(Self::NoteRef, IncrementalWitness<Node>)>, Self::Error> {
+        // database updates for each block are transactional
+        self.transactionally(|up| {
+            let db = up.wallet_db.inner.lock().unwrap();
+            // Insert the block into the database.
+            wallet_actions::insert_block(
+                &db,
+                block.block_height,
+                block.block_hash,
+                block.block_time,
+                block.commitment_tree,
+            )?;
+
+            let mut new_witnesses = vec![];
+            for tx in block.transactions {
+                let tx_row = wallet_actions::put_tx_meta(&db, tx, block.block_height)?;
+
+                // Mark notes as spent and remove them from the scanning cache
+                for spend in &tx.shielded_spends {
+                    wallet_actions::mark_spent(&db, tx_row, &spend.nf)?;
+                }
+
+                for output in &tx.shielded_outputs {
+                    let received_note_id = wallet_actions::put_received_note(&db, output, tx_row)?;
+
+                    // Save witness for note.
+                    new_witnesses.push((received_note_id, output.witness.clone()));
+                }
+            }
+
+            // Insert current new_witnesses into the database.
+            for (received_note_id, witness) in updated_witnesses.iter().chain(new_witnesses.iter())
+            {
+                if let NoteId::ReceivedNoteId(rnid) = *received_note_id {
+                    wallet_actions::insert_witness(&db, rnid, witness, block.block_height)?;
+                } else {
+                    return Err(SqliteClientError::InvalidNoteId);
+                }
+            }
+
+            // Prune the stored witnesses (we only expect rollbacks of at most 100 blocks).
+            let below_height = if block.block_height < BlockHeight::from(100) {
+                BlockHeight::from(0)
+            } else {
+                block.block_height - 100
+            };
+            wallet_actions::prune_witnesses(&db, below_height)?;
+
+            // Update now-expired transactions that didn't get mined.
+            wallet_actions::update_expired_notes(&db, block.block_height)?;
+
+            Ok(new_witnesses)
+        })
+    }
+
+    async fn store_received_tx(
+        &mut self,
+        received_tx: &ReceivedTransaction,
+    ) -> Result<Self::TxRef, Self::Error> {
+        self.transactionally(|up| {
+            let db = up.wallet_db.inner.lock().unwrap();
+            let tx_ref = wallet_actions::put_tx_data(&db, received_tx.tx, None)?;
+
+            for output in received_tx.outputs {
+                if output.outgoing {
+                    wallet_actions::put_sent_note(&db, output, tx_ref)?;
+                } else {
+                    wallet_actions::put_received_note(&db, output, tx_ref)?;
+                }
+            }
+
+            Ok(tx_ref)
+        })
```

Comment:
This is just an example for all implementations here
```rust
        let db = self.wallet_db.inner.lock().unwrap();
        let mut update_ops = db.get_update_ops()?;
        update_ops.store_received_tx(received_tx)
```

Reply:
this will require implementing `zcash_client_backend::data_api::WalletRead` for `DataConnStmtCacheAsync<P>` which is not something we want to do

Reply:
The following diff doesn't require `WalletRead` impl for `DataConnStmtCacheAsync`
```diff
--- a/zcash_client_sqlite/src/for_async/mod.rs
+++ b/zcash_client_sqlite/src/for_async/mod.rs
@@ -425,20 +425,10 @@ impl<P: consensus::Parameters + Send + Sync + 'static> WalletWrite for DataConnS
         &mut self,
         received_tx: &ReceivedTransaction,
     ) -> Result<Self::TxRef, Self::Error> {
-        self.transactionally(|up| {
-            let db = up.wallet_db.inner.lock().unwrap();
-            let tx_ref = wallet_actions::put_tx_data(&db, received_tx.tx, None)?;
-
-            for output in received_tx.outputs {
-                if output.outgoing {
-                    wallet_actions::put_sent_note(&db, output, tx_ref)?;
-                } else {
-                    wallet_actions::put_received_note(&db, output, tx_ref)?;
-                }
-            }
-
-            Ok(tx_ref)
-        })
+        use zcash_client_backend::data_api::WalletWrite;
+        let db = self.wallet_db.inner.lock().unwrap();
+        let mut update_ops = db.get_update_ops()?;
+        update_ops.store_received_tx(received_tx)
     }
```
@borngraced Could you recheck, please?

If it works, the implementation might be greatly simplified.

Reply:
yesss works. thank you so much!